### DISCUSSION TITLE
Bug with uppercase letters input: will fix be backported to Windows Powershell?

### DISCUSSION THREAD

**Discussion by shvchk** [PowerShell/PowerShell#17659] [Q&A]
There was a bug in PSReadLine which prevented capital letters input, see #10794, microsoft/terminal#7087 

It was probably fixed in PSReadLine 2.0 Beta 5, so installing PowerShell Core fixes the issue, as it is shipped with updated PSReadLine.

However, Windows PowerShell on fully updated Windows 10 is still affected by this bug. 
<details><summary>PS and PSRL versions:</summary>

```
> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      5.1.19041.1682
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.19041.1682
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1

> Get-Module PSReadline

ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Script     2.0.0      PSReadline                          {Get-PSReadLineKeyHandler, Get-PSReadLineOption, Remove-PS...
```
</details>

At the same time, Windows PowerShell on Windows 11 reports the same PSReadLine version, but does not seem to be affected by this bug (at least I haven't encountered it, but can't be sure since PSReadLine version is the same).
<details><summary>PS and PSRL versions:</summary>

```
> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      5.1.22000.653
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.22000.653
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1

> Get-Module PSReadline

ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Script     2.0.0      PSReadline                          {Get-PSReadLineKeyHandler, Get-PSReadLineOption, Remove-PS...
```
</details>

---

So, questions 🙂
- Is this bug really fixed in Windows PowerShell shipped with Windows 11?
- Will this fix be backported to Windows Powershell on Windows 10 and delivered via Windows Update?

**Comment by doctordns** [comment]
You can fix this yourself and just install the latest version of PSReadline

```powershell
Update-Module PSREADLINE
# or
Install-Module PSReadline
```

**Reply by shvchk** [reply]
Thanks, @doctordns. I know I can fix it manually, but users really shouldn't _have to_, we have Windows Update for that :wink:

Btw, `Update-Module` won't work for preinstalled module iirc (which is the case for PSReadline), `Install-Module` will work with `-Force`. Funny thing is, the nature of this bug will even prevent pasting this command to PS to fix it, since command contains uppercase letters :smile: User would need to input / paste it in lowercase, that will work.

**Reply by jborean93** [reply]
The fun thing is it will work. PowerShell is case insensitive when it comes to these things and luckily module names are also case insensitive . These will do the same thing

```ps
Update-Module -Name PSReadLine
update-module -name psreadline
UPDATE-MODULE -NAME PSREADLINE
```

What will be annoying with `Update-Module` and `PSReadLine` is the inability to unload and remove the dll's on Windows as they will have been loaded by the PowerShell process itself. The `Install-Module ... -Force` command you mentioned does allow it to be installed side by side and as mentioned above is case insensitive so can be written in any way you wish.

**Comment by kasini3000** [comment]
I think this is a batch operation and maintenance problem of win node machines.
To batch update the psreadline module, just do:

1Download the latest version of the psreadline module on the MASTER computer.

2 Copy the module directory to the NODE computer
C:\Program Files\WindowsPowerShell\Modules   #powershell v5

3 Copy the module directory to the node computer
C:\Program Files\PowerShell\Modules  #if powershell v7 installed

Obviously, the best way to do this is of course with the kasini3000 project I wrote. 
Some people don't want me to talk about this, if you are interested, search by yourself

**Reply by doctordns** [reply]
Using update-module or Install-module are built in and work fine. You do not need a third party module to deploy the update.

As anside, I do feel that update-module should work in all cases. But that is a separate issue

**Reply by kilasuit** [reply]
@kasini3000 I've asked you, on behalf of the PowerShell Working Group & the PowerShell team once before not to promote your project in issues/discussions in this repo. 
Please do not make me ask you a 3rd time & for what it's worth we don't mind you promoting your project & do want community members to do so when they are proud of their work, but please do it in the right way and not within issues/discussions as they aren't the best place for you to promote it. I would suggest that you perhaps look at seeing whether you can get a post written on [PowerShell.org](https://powershell.org/) to promote it.

**Reply by kasini3000** [reply]
>on behalf of the PowerShell Working Group & the PowerShell team

Who are you? What contributions have you made to the ps community? Why can you represent the team and the community?



>not to promote your project in issues/discussions

I am not clear about the rules for issues/discussions ,
where are the rules? Who are the rules restricting?
which one rule fits my situation?
Why don't you discuss addressing community style and leave it to you to decide?


>but please do it in the right way and not within issues/discussions

on this discussions: users has some win10/win11 pc,
do you know install/update module need admin rights
fast, batch, for all node, all powershell 5 and 7 (if install)
what is the best way?
Right click on the shortcut on powershell5 , and powershell7, on each pc?
@kilasuit I would like to hear your right way answer 

btw:
Would you like to talk about what contribution you have made? How did you help others and how many years did you persist?


-----

>You do not need a third party module to deploy the update.

yes.I think my way better.

### Do the powershell community prevent the use of third-party scripts and libraries to help users solve them problems on this discussions?

If others do the same, who is responsible for handling it ?


-----

my thoughts:
1 Although I am weak, I continue to use my efforts to solve problems for the powershell community.

2 Psreadline was not a team project before yesterday.
Some third-party modules are very close to the powershell core, even inseparable.
eg:
omi remoting, yaml module. also contains my framework.


3 I do a car ,powershell community do a engine,Both are important and should help each other.
I have a picture with a table comparing ansible, pipeline, salt,dare not post it.
Even I want to challenge Terraform with it.
you don't hope one powershell devops automation batch script framework like this？

If the powershell user base is a fire, do you want to add firewood? or pour water on the fire?

**Reply by kilasuit** [reply]
> Who are you? What contributions have you made to the ps community? Why can you represent the team and the community?

Did you look at my profile at all? Because I can only assume that you haven't and you really should. But to make it simple and clear, I am a [Microsoft MVP in Cloud and Datacenter Management](https://mvp.microsoft.com/en-us/publicprofile/5001821) and have been using PowerShell for almost a decade. I run and co-ordinate the UK PowerShell User Groups, as well as having helped organise PowerShell conferences like PSConfEU & PSDayUK as well as having taught hundreds of people how to use PowerShell and presented at user groups and conferences internationally since 2015. I've sat in the room with the PS team on many occasions, both physically and virtually and have helped drive a decent amount of change in the language over the years.

>I am not clear about the rules for issues/discussions ,
where are the rules? Who are the rules restricting?
which one rule fits my situation?

This is valuable feedback that I will ensure is made much clearer in future, so please believe me when I say that I really do thank you for raising this, after all this language cannot get better without input from a wide and diverse group of people, which I **hope** you will continue to be a part of, albeit in perhaps more in line with the spirit of the community and particularly this repo. I **definitely** do not want to discourage you from being a part of the community as I can see that you would be a valued member given the right guidance on how to best go about things.

> @kilasuit I would like to hear your right way answer

For this I haven't looked into it enough to be able to give one yet, however I may do so and then will come back with a suitable answer.

> Do the powershell community prevent the use of third-party scripts and libraries to help users solve them problems on this discussions?

No, however there is a right way to suggest that "hey I wrote this and I think it may help you" and a wrong way too. We want engagements and promotions like that to be conducted in the right way as not to annoy or put off other contributors like yourself in future.

> Although I am weak, I continue to use my efforts to solve problems for the powershell community.

Please don't see yourself as weak & please do continue in what you are doing to help the PowerShell Community.

> Psreadline was not a team project before yesterday.
Some third-party modules are very close to the powershell core, even inseparable.

Yes very true and one of my issues https://github.com/PowerShell/PowerShell/issues/1979 highlighted this somewhat.

> I do a car ,powershell community do a engine,Both are important and should help each other.
I have a picture with a table comparing ansible, pipeline, salt,dare not post it.
Even I want to challenge Terraform with it.
you don't hope one powershell devops automation batch script framework like this？

I'm not too clear on what you mean by this so perhaps you could expand a little

>If the powershell user base is a fire, do you want to add firewood? or pour water on the fire?

I will always gather and add firewood & use water on the fire in a teapot to make a nice cuppa tea, because there is nothing better than a warm cuppa by a hot fire.

I hope that clears things up and if you would like to chat with me 1 to 1 then I'd be more than happy to via whichever mean suits you best.

**Reply by kasini3000** [reply]
### The information is only for kilasuit

Sorry, I didn't see your profile. I'm not an mvp because that would expose my picture and name.

I call myself a "powershell preacher". I created the largest powershell discussion forum in China in 2011, 

which is the Tencent group, which is like icq,msn, and currently has less than 1,000 people.

For this problem, my original intention is to download from one host and copy it to all win, linux nodes through winrm,
Can I quote a script(or tech) from someone else's blog as an answer on discussions?
Are references to stack overflow (or else's) allowed?


From now on, I'd like to translate the introductory article for project "k", and then I'll go to Script guys blog to contribute.

I do a car ,powershell community do a car engine,The car I make is the best.

1 Have you ever expected PowerShell open source, free, cross platform, Devops tools.
One of the licenses of the tool is: as long as you have contributed to PowerShell, you can use it unlimited?
On my way to promote "k", I said to people over and over again: 
"It's enough to learn powershell for operation and maintenance personnel, you don't need yaml, you don't need ansible, you don't need python"

project "k". I think 2 years is the lead.
Maybe someone can do better? please tell me.

2 Based on the project "k", the project "k4t" has been completed.
which is a better container automatic operation and maintenance framework than kubernetes.
I fell powershell 10 times better than YAML
The project was not successful and I am looking forward to launching it in 5 years.

This is the craziest thing in my life. Currently, I'm learning golang and partially rewriting it in golang.
